src/Workspaces/Core/Portable/Rename/ConflictEngine/DeclarationConflictHelpers.cs

```diff
@@ -58,51 +61,43 @@ public bool Equals(List<ITypeSymbol> x, List<ITypeSymbol> y)
                 return x.SequenceEqual(y);
             }
 
-            public int GetHashCode(List<ITypeSymbol> obj)
+            public int GetHashCode(ImmutableArray<ITypeSymbol> obj)
             {
-                // This is a very simple GetHashCode implementation. Doing something "fancier" here
-                // isn't really worth it, since to compute a proper hash we'd end up walking all the
-                // ITypeSymbols anyways.
-                return obj.Count;
+                int hash = 13;
+                foreach (var type in obj)
+                {
+                    hash = Hash.Combine((int)type.TypeKind, hash);
+                }
+                return hash;
             }
         }
 
-        private static IEnumerable<List<ITypeSymbol>> GetAllSignatures(IMethodSymbol method, bool trimOptionalParameters)
+        private static ImmutableArray<ImmutableArray<ITypeSymbol>> GetAllSignatures(IMethodSymbol method, bool trimOptionalParameters)
         {
-            // First we'll construct the full signature. This consists of the types of the
-            // parameters, as well as the return type if it's a conversion operator
-            var signature = new List<ITypeSymbol>();
+            var resultBuilder = ArrayBuilder<ImmutableArray<ITypeSymbol>>.GetInstance();
+
+            var signatureBuilder = ArrayBuilder<ITypeSymbol>.GetInstance();
 
             if (method.MethodKind == MethodKind.Conversion)
             {
-                signature.Add(method.ReturnType);
+                signatureBuilder.Add(method.ReturnType);
             }
 
             foreach (var parameter in method.Parameters)
             {
-                signature.Add(parameter.Type);
-            }
-
-            yield return signature;
-
-            // In VB, a method effectively creates multiple signatures which are produced by
-            // chopping off each of the optional parameters on the end, last to first, per 4.1.1 of
-            // the spec.
-            if (trimOptionalParameters)
-            {
-                for (int i = method.Parameters.Length - 1; i >= 0; i--)
+                // In VB, a method effectively creates multiple signatures which are produced by
+                // chopping off each of the optional parameters on the end, last to first, per 4.1.1 of
+                // the spec.
+                if (trimOptionalParameters && parameter.IsOptional)
                 {
-                    if (method.Parameters[i].IsOptional)
-                    {
-                        signature = signature.GetRange(0, signature.Count - 1);
```

Comment:
Actually, this does the same amount of copying.  I maintain that it's clearer though.